### PR TITLE
Add gprofiler2 module

### PR DESCRIPTION
**Pull request by WackerO**
This PR adds a [gprofiler2](https://cran.r-project.org/web/packages/gprofiler2/vignettes/gprofiler2.html) module for pathway enrichment analysis

## PR checklist

Closes #XXX 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by WackerO** [comment]
Umm so I'm kind of out of ideas about that conda fail...When I test the module locally, all three profiles pass, so I assume it must be some minor package versioning difference maybe. I checked the different files out (managed to reproduce the fail on gitpod), but the resulting plots look identical to my eye, the difference is only obvious from the checksums.

Should I just remove the md5sums or can we ignore the conda error? I searched the nf-core slack and found a bunch of people who also had trouble with conda tests producing different md5sums vs docker/singu (even though those were usually .gz or .bam etc., not images).

File: modules/nf-core/gprofiler2/gost/meta.yml
Comment (vagkaratzas) [inline modules/nf-core/gprofiler2/gost/meta.yml:114]: add meta in the output as well

File: modules/nf-core/gprofiler2/gost/meta.yml
Comment (vagkaratzas) [inline modules/nf-core/gprofiler2/gost/meta.yml:141]: add file patterns for the rest of the outputs as well (as in versions)

File: modules/nf-core/gprofiler2/gost/main.nf
Comment (vagkaratzas) [inline modules/nf-core/gprofiler2/gost/main.nf:5]: I am not sure how double conda envs are standardised in nf-core, so somebody else should have to check if containers are OK here. And/or if we need to move these to the environment.yml file in the main folder as well.
Would need to change this line here to:
conda "${moduleDir}/environment.yml"

and then have an environment.yml file:
channels:
  - conda-forge
  - bioconda
  - defaults
dependencies:
  - conda-forge::r-ggplot2=3.4.3
  - conda-forge::r-gprofiler2=0.2.2

File: modules/nf-core/gprofiler2/gost/main.nf
Comment (vagkaratzas) [inline modules/nf-core/gprofiler2/gost/main.nf:29]: I am also unsure about template calling in nf-core, so again, would need a second reviewer on this.

File: modules/nf-core/gprofiler2/gost/meta.yml
Comment (vagkaratzas) [inline modules/nf-core/gprofiler2/gost/meta.yml:12]: probably have this as the homepage instead? https://biit.cs.ut.ee/gprofiler/page/r
can still use the other link as the dev url below I think

File: modules/nf-core/gprofiler2/gost/meta.yml
Comment (vagkaratzas) [inline modules/nf-core/gprofiler2/gost/meta.yml:28]: Someone else please confirm that we can write in meta arguments that are not in the main.nf inputs

**Review by vagkaratzas — COMMENTED** [review]
Seems to do the job. Will just need someone else to clarify some details before approval.

**Comment by sateeshperi** [comment]
all valid points from @vagkaratzas 

@WackerO kindly address those please and let us know if you have any questions. Thanks

File: modules/nf-core/gprofiler2/gost/main.nf
Comment (WackerO) [reply modules/nf-core/gprofiler2/gost/main.nf:5]: Ou, I did not notice nf-core had switched to an env.yml! Done!

File: modules/nf-core/gprofiler2/gost/meta.yml
Comment (WackerO) [reply modules/nf-core/gprofiler2/gost/meta.yml:28]: @sateeshperi Hey, should I then remove from the meta all arguments that are not channels?

File: modules/nf-core/gprofiler2/gost/main.nf
Comment (WackerO) [reply modules/nf-core/gprofiler2/gost/main.nf:29]: This is done in some other modules as well, like here :)
https://github.com/nf-core/modules/blob/master/modules/nf-core/deseq2/differential/main.nf

File: modules/nf-core/gprofiler2/gost/templates/gprofiler2_gost.R
Comment (pinin4fjords) [inline modules/nf-core/gprofiler2/gost/templates/gprofiler2_gost.R:227]:
```suggestion
contrast_name <- paste(opt\$contrast_variable, opt\$reference_level, opt\$target_level, sep = '_')
if (!is.null(opt\$blocking_variables)) {
    blocking_variables <- paste(make.names(unlist(strsplit(opt\$blocking_variables, split = ','))), collapse = '_')
    output_prefix <- paste(contrast_name, blocking_variables, sep= '_')
}
output_prefix <- paste('gprofiler2', contrast_name, sep = '.')
```

File: modules/nf-core/gprofiler2/gost/templates/gprofiler2_gost.R
Comment (pinin4fjords) [inline modules/nf-core/gprofiler2/gost/templates/gprofiler2_gost.R:255]: Don't call out unnecessarily, use download.file()

**Review by pinin4fjords — COMMENTED** [review]
I'm partly echoing existing feedback.

I also really hope that we can keep these R template packages to wrappers around the target functions. If you're adding your own stuff (beyond argument / file handling etc), the code should really be living in external repos (in my opinion, others may differ).

File: modules/nf-core/gprofiler2/gost/meta.yml
Comment (WackerO) [reply modules/nf-core/gprofiler2/gost/meta.yml:77]: Ah okay, I was not sure about optional files, thanks!

**Comment by WackerO** [comment]
> I'm partly echoing existing feedback.
> 
> I also really hope that we can keep these R template packages to wrappers around the target functions. If you're adding your own stuff (beyond argument / file handling etc), the code should really be living in external repos (in my opinion, others may differ).

Hmm okay. Maybe someone else can give input on this as well? I understood that templates are a simple way to add some external code to a module script without cramming all of it into the main.nf itself (also from the bytesize talk about this topic).

**Comment by pinin4fjords** [comment]
> Hmm okay. Maybe someone else can give input on this as well? I understood that templates are a simple way to add some external code to a module script without cramming all of it into the main.nf itself (also from the bytesize talk about this topic).

Oh yes, going in the main.nf would be worse. But something external would be better. 

As I say, I won't fight you on this point anyway, just wanted to highlight it as something to aim not to do more of. When I get round to it I'm going to move the code in the  deseq et al modules to external packages.

**Comment by WackerO** [comment]
@sateeshperi I think I went over all of your and Jon's requested changes and applied/implemented them :)

**Comment by WackerO** [comment]
The two failing tests are because of another module, maybe that can be overlooked...

**Review by vagkaratzas — COMMENTED** [review]
Are the inputs separated into three different tuples thematically?
tuple val(meta), val(contrast_variable), val(reference), val(target)
tuple val(meta1), path(de_file)
tuple val(organism), val(sources)

If different meta are needed for different tuples, be sure to include them in the meta.yml file. Other than that, and the fact that there is too much code in the .R file (other than just calling the existing gprofiler functions), I think it should be OK to merge.

**Comment by WackerO** [comment]
> gmt_file input, not in the meta.yml reference input, named reference_level in the meta.yml, make them the same same for target / target_level
> 
> filtered_gmt output not in meta.yml session_info output not in meta.yml

Hmm, then maybe I understood a previous comment incorrectly; I thought these things do not go into the meta.yml. But I'll re-add them

File: modules/nf-core/gprofiler2/gost/meta.yml
Comment (vagkaratzas) [inline modules/nf-core/gprofiler2/gost/meta.yml:55]: Don't need curly braces in the pattern when only one option

**Review by vagkaratzas — APPROVED** [review]
Goodie! Just a minor comment again on the meta and I think you can merge

File: modules/nf-core/gprofiler2/gost/meta.yml
Comment (WackerO) [reply modules/nf-core/gprofiler2/gost/meta.yml:55]: Oops, you are right!

**Comment by WackerO** [comment]
Thanks for the review!

**Pull request merged**
